File: src/cpp/bvh.cpp
#include "bvh.h"

bool BVHNode::hit(const Ray& ray, const Interval& interval, HitRecord& record) const {
    if (!hit_bbox(this->bbox, ray, interval)) {
        record.hit = false;
        return false;
    }

    bool hit_left  = false;
    bool hit_right = false;

    if (this->left != nullptr) {
        hit_left = this->left.get()->hit(ray, interval, record);
    }

    if (this->right != nullptr) {
        hit_right = this->right.get()->hit(
            ray, Interval(interval.lo, hit_left ? record.t : interval.hi), record
        );
    }

    record.hit = hit_left || hit_right;

    return record.hit;
}

File: src/python/modules/bvh.py
from functools import singledispatch
from .aabb import hit_aabb
from .models import LIB_SCOPE, Ray, Interval, HitRecord, wrapper_dispatch, BVHNode


@singledispatch
def hit(node, ray: Ray, interval: Interval, record: HitRecord) -> HitRecord:
    if not hit_aabb(wrapper_dispatch(node.bbox), ray, interval):
        return record

    left = hit(wrapper_dispatch(node.left), ray, interval, record)
    right = hit(
        wrapper_dispatch(node.right), ray, Interval(
            interval.lo, record.t if left.hit else interval.hi), record
    )

    return left if left.hit else right
